### PR TITLE
Support proxies for PeerTube

### LINKED ISSUE #4202: Support proxies for HTTP requests used by PeerTube
**Describe the current behavior**
After updating to 3.2.x, it fails to retrieve the plugin list.

image

I use peertube behind corporate proxy, and in 3.1.x, I was able to search and install plugins by setting the environment variables 
`http_proxy`  and `https_proxy` .

I guess it because of http library.
Since v3.2.x, http library is changed from [request](https://github.com/request/request) to [Got](https://www.npmjs.com/package/got).

[request](https://github.com/request/request) support http_proxy, https_proxy with environment variables, but [Got](https://www.npmjs.com/package/got) needs to use `tunnel` or `hpagent` library for supporing proxy.

**Additional information**

* PeerTube instance:
  * URL: private
  * version: 3.2.1
  * NodeJS version: 12.22.1
  * Ffmpeg version: 4.1.6-1~deb10u1

* console log:
```
error[2021/6/21 17:36:29] Cannot list available plugins from index https://packages.joinpeertube.org/api/v1/plugins.
{
  "err": {
    "stack": "RequestError: connect ETIMEDOUT 178.63.240.150:443\n    at ClientRequest.<anonymous> (/app/node_modules/got/dist/source/core/index.js:956:111)\n    at Object.onceWrapper (events.js:421:26)\n    at ClientRequest.emit (events.js:326:22)\n    at ClientRequest.EventEmitter.emit (domain.js:483:12)\n    at ClientRequest.origin.emit (/app/node_modules/@szmarczak/http-timer/dist/source/index.js:39:20)\n    at TLSSocket.socketErrorListener (_http_client.js:427:9)\n    at TLSSocket.emit (events.js:314:20)\n    at TLSSocket.EventEmitter.emit (domain.js:483:12)\n    at emitErrorNT (internal/streams/destroy.js:92:8)\n    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)\n    at processTicksAndRejections (internal/process/task_queues.js:84:21)\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1144:16)",
    "message": "connect ETIMEDOUT 178.63.240.150:443",
    "name": "RequestError"
  }
}
```

### PR DESCRIPTION
**Pull request by smilekison**
## Description

This pull request is made to solve the issue in response to #4202. This functionality is useful for the corporate office who wish to run PeerTube behind the Proxy environment. With this update, the PeerTube checks for the availability of proxy addresses in the OS Environment. The process in which PeerTube runs is as follow: 

1. Corporate offices can run PeerTube behind Proxy by providing HTTPS_PROXY or HTTP_PROXY (case sensitive: must be UPPERCASE & in URL format) in the OS environment variable.

2. If the Proxy variable is not set in the OS Environment, the PeerTube runs without proxy.

We have already shared the detailed version of Architecture and the code in the respective issue and we got approval to create a pull request by the owner of PeerTube. We have tested the application with test coverage and the respective documentation can be found in the following link.
 
https://docs.google.com/document/d/1Hg7fKbGnAqPmZsfbzvozH36LjGkmnqITK0oAxntBv4o/edit?usp=sharing
## Related issues
Fixes #4202

## Has this been tested?

- [x] 👍 yes, I added tests to the test suite
- [ ] 💭 no, because this PR is a draft and still needs work
- [ ] 🙅 no, because this PR does not update server code
- [ ] 🙋 no, because I need help

Co-authored-by: Sijan Bhattarai <bhattaraisijan@outlook.com>
Co-authored-by: Julesh Shrestha <juliusshrestha29@gmail.com>
Co-authored-by: Durgaraj Karki <raazkarkee405@gmail.com>

### CODE REVIEW

File: server/helpers/requests.ts [hunk 1]
```diff
@@ -28,7 +31,21 @@ type PeerTubeRequestOptions = {
   jsonResponse?: boolean
 } & Pick<GotOptions, 'headers' | 'json' | 'method' | 'searchParams'>
 
-const peertubeGot = got.extend({
+
+let peertubeGot: Got
+if (proxy()) {
+  console.log('With Proxy Address')
+peertubeGot = got.extend({
+  agent: {
```
Comment (Chocobozzz) [inline server/helpers/requests.ts:39]: Don't use an `if/else`: you duplicate too much code

Instead, inject the agent property: 

```

...getAgentIfNeeded(),

headers: {
...
```

and 

```
function getAgentIfNeeded () {
  if (!isProxyEnabled()) return {}

  return { 
    agent: {
      ....
    }
  }
}
```

File: server/middlewares/proxy.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+const envs = require("env-smart")
```
Comment (Chocobozzz) [inline server/middlewares/proxy.ts:1]: You don't need `env-smart`, just use `process.env.HTTP_PROXY` and `process.env.HTTPS_PROXY`

File: server/middlewares/proxy.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    proxyAddress = OS_HTTPS_PROXY
+  }
+  /* istanbul ignore else */
+  if (typeof OS_HTTP_PROXY !== "undefined") {
+    console.log("OS HTTP_PROXY Environment variable :", `${OS_HTTP_PROXY}`)
+    proxyAddress = OS_HTTP_PROXY
+  }
+  return proxyAddress
+}
+module.exports = proxy
```
Comment (Chocobozzz) [inline server/middlewares/proxy.ts:19]: Put `proxy.ts` in `helpers/` instead, this is not a middleware

File: server/middlewares/proxy.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+const envs = require("env-smart")
+// read environment variable
+const OS_HTTP_PROXY = envs.load().HTTP_PROXY
+const OS_HTTPS_PROXY = envs.load().HTTPS_PROXY
+const proxy = () => {
```
Comment (Chocobozzz) [inline server/middlewares/proxy.ts:5]: `function getProxy () {`

File: server/middlewares/proxy.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  let proxyAddress = null
+  /* istanbul ignore else */
+  if (typeof OS_HTTPS_PROXY !== "undefined") {
+    console.log("OS HTTPS_PROXY Environment variable :", `${OS_HTTPS_PROXY}`)
```
Comment (Chocobozzz) [inline server/middlewares/proxy.ts:9]: Use `logger.debug` instead of `console.log`

File: server/middlewares/proxy.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
+const envs = require("env-smart")
+// read environment variable
+const OS_HTTP_PROXY = envs.load().HTTP_PROXY
+const OS_HTTPS_PROXY = envs.load().HTTPS_PROXY
+const proxy = () => {
+  let proxyAddress = null
+  /* istanbul ignore else */
```
Comment (Chocobozzz) [inline server/middlewares/proxy.ts:7]: ?

File: support/docker/production/Dockerfile.buster [hunk 7]
```diff
@@ -35,6 +35,9 @@ RUN chown -R peertube:peertube /data /config
 ENV NODE_ENV production
 ENV NODE_CONFIG_DIR /config
 
+ENV HTTP_PROXY=$HTTP_PROXY
```
Comment (Chocobozzz) [inline support/docker/production/Dockerfile.buster:38]: We don't need this

File: support/docker/production/docker-compose.yml [hunk 8]
```diff
@@ -51,6 +51,11 @@ services:
     networks:
       default:
         ipv4_address: 172.18.0.42
+    
+    environment:
```
Comment (Chocobozzz) [inline support/docker/production/docker-compose.yml:55]: We don't need this, inject them in `.env` file instead

**Review by Chocobozzz — CHANGES_REQUESTED** [review]

File: server/helpers/requests.ts (hunk 1 above)
Comment (smilekison) [reply server/helpers/requests.ts]: We have removed the redundancy and made the code smaller as you suggested.

File: server/middlewares/proxy.ts (hunk 2 above)
Comment (smilekison) [reply server/middlewares/proxy.ts]: We have used the process.env.HTTP_PROXY & process.env.HTTPS_PROXY with this changes.

File: server/middlewares/proxy.ts (hunk 3 above)
Comment (smilekison) [reply server/middlewares/proxy.ts]: We have moved the proxy.ts file from middlewares to helpers.

File: server/middlewares/proxy.ts (hunk 4 above)
Comment (smilekison) [reply server/middlewares/proxy.ts]: We have properly defined the function as you suggested.

File: server/middlewares/proxy.ts (hunk 5 above)
Comment (smilekison) [reply server/middlewares/proxy.ts]: we have used logger.debug instead of console.log as you suggested.

File: server/middlewares/proxy.ts (hunk 6 above)
Comment (smilekison) [reply server/middlewares/proxy.ts]: we have removed the redundancy.

File: support/docker/production/Dockerfile.buster (hunk 7 above)
Comment (smilekison) [reply support/docker/production/Dockerfile.buster]: We removed the ENV HTTP_PROXY=$HTTP_PROXY from Dockerfile.buster

File: support/docker/production/docker-compose.yml (hunk 8 above)
Comment (smilekison) [reply support/docker/production/docker-compose.yml]: we have removed environment from docker-compose.yml file.

**Comment by Chocobozzz** [comment]
Thanks!

**Pull request merged**